internal/provider/data_source_certificate.go

```diff
@@ -139,3 +160,33 @@ func dataSourceCertificateRead(d *schema.ResourceData, _ interface{}) error {
 
 	return nil
 }
+
+func fetchPeerCertificatesViaTLS(targetURL *url.URL, shouldVerifyChain bool) ([]*x509.Certificate, error) {
+	conn, err := tls.Dial("tcp", targetURL.Host, &tls.Config{
+		InsecureSkipVerify: !shouldVerifyChain,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	return conn.ConnectionState().PeerCertificates, nil
+}
+
+func fetchPeerCertificatesViaHTTPS(targetURL *url.URL, shouldVerifyChain bool, config *providerConfig) ([]*x509.Certificate, error) {
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: !shouldVerifyChain,
+			},
+			Proxy: config.proxyForRequestFunc(),
+		},
+	}
+	resp, err := client.Get(targetURL.String())
```

Comment:
Previously the data source would only open a TLS connection and not attempt to finish an actual HTTP call (in this new code the `GET` verb) or read the response body from the URL. A few questions:

* Can we avoid downloading the response body? e.g. should this execute a `HEAD` request or might a custom request handler function help here?
* Do we need to do anything to avoid new errors from switching to net/http?

Reply:
About using `HEAD` vs `GET` comes out of my experience that out there there are still HTTP servers that don't implement `HEAD` by default. In the past, where I had to do this kind of stuff, I'd even have logic to try `HEAD-then-GET`. I'm frankly unsure what's the best option here.

About the full HTTP call, I'm also unsure. It's out of this uncertainty that I introduced the support for `tls://` vs `https:/` schemes: my thinking was, until now we enforced `https:/` but actually never executed HTTP in the first place - just opened a socket.

So, I thought that adding `tls://` was going to be an easy way to suggest a "if you want to keep the previous behaviour...".

I'm frankly unsure what's the best strategy here. I'm tempted to suggest to make this PR a breaking change, because the only alternative I see is a `.mode` attribute that allows to chose the way the resource works. And I don't like it much.

Reply:
Just to capture some of our out-of-band conversation (please correct me if I misspeak anywhere!):

1. We are going to try and implement this so the HTTP client is avoided, continuing to use a direct TLS connection instead, unless a HTTP proxy has been configured (currently via the provider `proxy` configuration block)
2. Since IANA does not have a URI scheme for strictly a TCP network socket connection, we can use a synthetic `tls://` scheme (or potentially use _no_ scheme).
3. When using a HTTP proxied client, it will first attempt a `HEAD` call, then fallback to a `GET` call in case of a proxy that does not support `HEAD` and blocks the subsequent HTTP connection.

Not discussed, but thinking out loud: I'm presuming we can capture 405 Method Not Allowed or 501 Not Implemented HTTP status codes for the HTTP proxied client fallback. Another thing I'm not sure about is whether we will need special 407 Proxy Authentication Required handling anywhere.

Reply:
Actually, while I want to handle `HEAD` then `GET`, I think we can't give lots of weight to the actual HTTP response: the TLS negotiation happens much earlier then any HTTP Protocol byte has moved.

So, I think it makes sense to handle error response from an `http.Get()` call, but if the `reps.TLS.PeerCertificates` are there, this provider shouldn't do much more than process those.

I'll show you what I mean in the next commits I push.